### ISSUE TITLE
Massive memory allocations and reduced performance in GradientTape for loop and Optimizer when using jit_compile

### ISSUE THREAD

**Issue by CodingRays** [tensorflow/tensorflow#50198]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Fedora 34 Workstation Edition
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): v2.5.0-rc2-14-gfcdf6593470 2.5.0-rc3
- Python version: 3.9.5
- GPU model and memory: Running on CPU
(If it is important the CPU is a 3900X and has 32GB of RAM)

I recently ran into issues with a custom training loop and `jit_compile` so i ran some small tests and found massive performance issues when enabling `jit_compile`. 

The code is a simple for loop inside a `tf.GradientTape` context (see below) where i accumulate a loss value. (Yes in this case i could batch the calls but that sadly isn't possible in the original issue i was talking about earlier)

In my test i ran the exact same training code once with `jit_compile` enabled and once with just using `tf.function`. I found that when enabling `jit_compile` the code runs approximately 1.7 times slower while consuming 12GB of memory as opposed to 1.4GB without `jit_compile`. I do not know if the slowdown is a direct result of the memory usage or due to other issues.
If i remove the `tf.GradientTape` the memory issue vanishes however the speed issue still persists. I also tried specifically setting the watched variables with no change.
If i leave the gradient tape in and just remove the call to the optimizer the memory usage drops to 8.8GB however performance is now over 25 times worse (16.5s instead of 0.6s). Which seems to suggest that both the optimizer and the GradientTape have problems.
Lastly i also tried setting a constant value directly in the range inside the for loop which also did not change the memory or performance issues.

I would expect the jit compiled code to at least have the same memory footprint as the not compiled version.

**The code:**
```
import tensorflow as tf
import timeit

model = tf.keras.Sequential([
    tf.keras.layers.InputLayer((2, )),
    tf.keras.layers.Dense(2048, activation='elu'),
    tf.keras.layers.Dense(2048, activation='elu'),
    tf.keras.layers.Dense(1)
])

opt = tf.keras.optimizers.Adam()


@tf.function
def normal_func(x, y, k):
    print("Tracing normal")
    with tf.GradientTape() as tape:
        loss = tf.constant((0, ), dtype=tf.float32)
        for i in tf.range(k):
            geny = model(tf.expand_dims(x[i], axis=0), training=True)[0]
            loss = loss + tf.square(y[i] - geny)

    grads = tape.gradient(loss, model.trainable_variables)
    opt.apply_gradients(zip(grads, model.trainable_variables))
    return loss


@tf.function(jit_compile=True)
def jit_func(x, y, k):
    print("Tracing jit")
    with tf.GradientTape() as tape:
        loss = tf.constant((0, ), dtype=tf.float32)
        for i in tf.range(k):
            geny = model(tf.expand_dims(x[i], axis=0), training=True)[0]
            loss = loss + tf.square(y[i] - geny)

    grads = tape.gradient(loss, model.trainable_variables)
    opt.apply_gradients(zip(grads, model.trainable_variables))
    return loss


x_ = tf.reshape(tf.range(0, 1024, dtype=tf.float32), shape=(512, 2))
y_ = tf.reshape(tf.range(0, 512, dtype=tf.float32), shape=(512, ))

k_ = tf.constant(8)
k2_ = tf.constant(256)
k3_ = tf.constant(512)

# Sanity tests for retracing
normal_func(x_, y_, k_)
normal_func(x_, y_, k2_)
normal_func(x_, y_, k3_)

jit_func(x_, y_, k_)
jit_func(x_, y_, k2_)
jit_func(x_, y_, k3_)

# Actual performance tests
print('Runtime without jit_compile:')
print(timeit.timeit(lambda: normal_func(x_, y_, k_), number=10))
print(timeit.timeit(lambda: normal_func(x_, y_, k2_), number=10))
print(timeit.timeit(lambda: normal_func(x_, y_, k3_), number=10))

print('\nRuntime with jit_compile:')
print(timeit.timeit(lambda: jit_func(x_, y_, k_), number=10))
print(timeit.timeit(lambda: jit_func(x_, y_, k2_), number=10))
print(timeit.timeit(lambda: jit_func(x_, y_, k3_), number=10))
```

The output when running on my machine:
[output.log](https://github.com/tensorflow/tensorflow/files/6631656/output.log)

**Comment by tilakrayal** [comment]
@rmothukuru ,
I was able to reproduce the issue in v2.5 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/88c56e984d474e69a27518fd1593ee60/50198.ipynb).

**Comment by rmothukuru** [comment]
@CodingRays,
I suspect the line `loss = loss + tf.square(y[i] - geny)` is the culprit. 

Can you please explain why do you want to add the **`Loss`** that way as it is not mentioned to do so in the [documentation of Custom Training Loop](https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch#using_the_gradienttape_a_first_end-to-end_example)?

**Comment by CodingRays** [comment]
Im not sure what you mean with not mentioned in the documentation. I dont see this situation mentioned in the docs. I just need some way to accumulate the losses over the iterations of the for loop while still being inside the gradient tape. In the docs the gradient tape is always inside the for loop.
If it helps for context the original problem i had was with updating a lstm based ddpg model and there was some processing i had to do in between the timesteps which is why i wasnt able to batch it.

**Comment by cheshire** [comment]
Note that on GPU performance numbers are as expected:

```
Runtime without jit_compile:
0.053597197867929935
1.1324485160876065
2.3468134391587228

Runtime with jit_compile:
0.039613737957552075
0.7868384378962219
1.581948912004009
```

**Comment by sushreebarsa** [comment]
@CodingRays Could you please confirm if the issue still persists in latest TF version ?
Thank you!

**Comment by CodingRays** [comment]
Yes it does. Memory still has the same issue. Performance however seems to have gotten significantly worse. I have tested 2.10rc3, 2.9.0 and 2.8.0 (because that is the oldest one i could find on pip). My system is still the same except for an upgraded gpu and os (though that shouldnt matter for this test as its ran on the cpu).

2.10rc3:
```
Runtime without jit_compile:
0.3214342410001336
8.442433143999551
16.870503825000014

Runtime with jit_compile:
0.9527878999997483
22.74526109300041
45.134182935000354
```

2.9.0:
```
Runtime without jit_compile:
0.31529049999971903
8.302056960999835
16.574833343000137

Runtime with jit_compile:
0.9515738340005555
23.05654790999961
45.77620842899978
```

2.8.0:
```
Runtime without jit_compile:
0.32451256300009845
8.366467630999978
16.648677719999796

Runtime with jit_compile:
0.6396097030001329
12.721525485000711
25.195542323000154
```
So this is a ~2x reduction from 2.8.0 which itself is a ~50% reduction from the old results using 2.5

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
